doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,359 @@
+2. `Q_expiration` will contain the temporal predicate `mz_now() <= upper_ts` which ensures each
+   record is retracted at the upper bound of its validity period.
+3. `Q_final` will contain the temporal predicate `lower_ts <= mz_now()` and a projection that hides
+   the last two columns, which contain the two timestamps of the validity period. This will ensure
+   that records are only inserted at the lower bound of their validity period and that `Q_expr` will
+   see rows with the correct schema.
+
+Given this decomposition we can now define the ingestion as:
+
+```
+C = Q_final(PERSIST(Q_expiration(Q_safemfp(RAW))))
+```
+
+Note that in the equation above `Q_final` is applied on the *output* of durably persisting
+persisting the rest of the ingestion. This MFP fragment will be deferred and applied on the read
+side when a dataflow uses the collection `C`.
+
+With these equations we have accomplished the goal of doing as much work as possible before the
+first persist operator while requiring no past state from `RAW` to resume. To see why that is the
+case, let's describe the state needed by the `Q_expiration` operator. The state of `Q_expiration` at
+time `t` is the set of records `(row, t_record, diff)` for which `t_record <= t && t <
+upper_ts(Q_mfp(row))`. In other words, for every timestamp it needs to remember all past records
+that expire after that timestamp.
+
+If we now look at the output of `Q_safemfp` we see that it converts a `(row, t_record, diff)` update
+to `((row, lower_ts, upper_ts), t_record, diff)` update. Additionally, `Q_expiration` will transform
+each `((row, lower_ts, upper_ts), t_record, diff)` update into two:
+
+1. `((row, lower_ts, upper_ts), t_record, diff)`
+2. `((row, lower_ts, upper_ts), upper_ts, -diff)`
+
+Given the above we can see that at the beginning of an ingestion execution `Q_expiration` can
+recover its in-memory state at time `t` by taking a snapshot of `C_intermediate` (not `RAW`) at time
+`t`, which is available because it is stored in persist! This is because any record that may expire
+in the future is maintained until the end of its validity period and `Q_mfp` has preserved the
+previously computed timestamps. This technique is similar to how the `UPSERT` envelope recovers its
+state from its output.
+
+### Planner/Optimizer stabilization
+
+When accepting a `CREATE SOURCE` statement with a `TRANSFORM USING` clause we are at that moment
+making a promise that that transformation will always be compatible with the storage execution
+model.
+
+In practice, since we have limited the accepted queries to only those that fully compile to an MFP,
+this means that the planner/optimizer must maintain the following invariant:
+
+> If a query `Q` fully optimizes to an MFP at some version `n` of Materialize, then it will continue
+> to plan and optimize to an MFP for all versions `n' >= n`. The MFPs are allowed to vary across
+> versions to allow for new optimizations to be taken advantage of.
```

Comment:
I think we should also say that the optimizer is only responsible for upholding this guarantee for `Q` that are _actually live in production_. I think that eliminates some stress. We don't need to do this for the universe of all possible `Q`, just the `Q` that customers have actually typed.

Reply:
Agreed! And as part of the implementation of this feature we should come up with a way to verify that before a release.

Reply:
This approach seems to have less impact to what has been discussed in the past! Still, I think we have some work to do to stabilize MFPs. The proposal requires that once a query optimized to an MFP, it'll do so for all times. I think this isn't generally true (but might be desirable!) because we don't have the confidence in the optimizer to make the choice, and because there's a degree of freedom to put things in an MFP or render it in a different way.

Consider rendering `IN` expressions. We can either express `x IN ('a', 'b')` as `x == 'a' OR x == 'b'`, or use a join with a constant arrangement to produce the same result. This shows that it's not clear that any MFP will stay an MFP. From a performance point of view we should use the MFP, but currently we don't make any promises like this, and we don't have the necessary confidence that our optimizer makes the right decisions in all cases. It should, but who knows.

What I can imagine to solve this problem is to allow-list expressions that we guarantee will result in an MFP. For example, we could stabilize `mz_now`, interval types, column references, casts, maybe JSON parsing, to enable some of the above examples. Sub-selects are tricky, because it requires quite a bit of machinery in the optimizer to determine that they can be inlined. This way, we could incrementally stabilize parts of the plan without fixing the status-quo.

I'm not sure how we could implement this, though. Someone who has more insight would need to judge if this approach could work or is a non-starter.

Reply:
> We can either express `x IN ('a', 'b')` as `x == 'a' OR x == 'b'`, or use a join with a constant arrangement to produce the same result.

Hmm, interesting example! I think this one is safe though, because we only consider making it a join if `x` comes from a Get to a GlobalId for which there is an existing index, which is impossible in the context of TRANSFORM USING. (A related future optimization, which I'm planning to implement at some point, is if we have `x == 1 OR x == 2 OR ... OR x == 1000`, and there is no index, then transform this boolean expression into a call to a special function that holds a BTreeSet of the 1, 2, ..., 1000 internally, and it probes that BTreeSet instead of evaluating an 1000-argument OR. But this optimization will also be fine, because it will change from one MFP to an other MFP.)

> What I can imagine to solve this problem is to allow-list expressions that we guarantee will result in an MFP. 

Let's say we operate on the HIR level:

In `HirScalarExpr` I would ban a few things:
- I would ban subqueries anywhere other than in the FROM clause. (The FROM clause exception is used by the "Decode JSON" example of the design doc.)
- I would ban `Windowing`.

In `HirRelationExpr` we could have an allow-list:
- Get
- Map, Filter, Project
- CallTable is probably needed, because it is necessary for JSON stuff, right? We should still check this a bit though, whether it might cause any problems.
- (Constant is not needed, right?)

My aim with the above is to only allow a linear chain of transformations, in the sense of not having any relation expression used more than once (no `HirRelationExpr::Let` or `LetRec`), and not having any relation expression with more than one input (e.g., Join, Union, or certain subquery patterns). This would prevent a lot of complex parts of the optimizer from coming into the picture at all, e.g., anything about joins, reduces, union cancellation, decorrelation.

Reply:
What I'd love to understand better is the timelines and amount of work involved in order for compute to feel comfortable stabilizing the as-written MFP guarantee.

I'd also like to check on whether the compute team thinks it's desirable goal! Is making this stability guarantee something that the compute team would _want_ to do, given enough time and testing etc, or would it be frustrating to optimization efforts to have this constraint?

With respect to the HIR allowlist, I think it's an idea that's worth exploring, but I have some concerns. Many simple transformations require heinous SQL queries. The JSON queries Gabor mentions are a great example. Those get planned into something very ugly but can often reduce down into an MFP. It seems like getting the allowlist right might require constant tinkering. I can imagine we're constantly updating the allowlist to allow for new, valid SQL patterns—or even struggling to come up with an allowlist that is neither too permissive nor too restrictive. On the other hand: I might be overestimating how hard it is to write this allowlist.

Basically my question is: if the compute team thinks the goal of MFP stability is desirable, are we better off putting our effort behind making that possible (mostly testing, I think), vs trying to build the allowlist?

Reply:
> Basically my question is: if the compute team thinks the goal of MFP stability is desirable, are we better off putting our effort behind making that possible (mostly testing, I think), vs trying to build the allowlist?

It’s true that the goal of MFP stability is desirable in any case, even without TRANSFORM USING, but TRANSFORM USING would make it much more strict, since then it will have to 100% work for those cases that are active at a customer. Btw. [the recent effort](https://github.com/MaterializeInc/materialize/pull/17245) to add normalizations to the optimizer goes in this direction, but it’s not making any hard guarantees, just reduces the occurrence of regressions.

After thinking about this more, I’m leaning towards no allowlist. In this case, testing optimizer changes on all TRANSFORM USING queries in production catalogs would be an absolute must in my opinion. (As a side note, even without TRANSFORM USING, it would be great to have tests that try whether all customer catalogs can boot with a new Materialize version.) This would eliminate a big portion of the risk. Two remaining risks that I can see:
1. (raised by Moritz) A customer might create a problematic TRANSFORM USING query between our last test run and the new version going live. This dangerous time interval would be about a day if we run these tests nightly, and could be made even less by an extra last-minute-before-maintenance-window test if this is a big worry. Even with 1 day, the risk is around 2 orders of magnitude less than without the testing (if we assume that a customer has hundreds of days to write a problematic query that _does_ get caught by the test), which might be acceptable. Anyhow, if this misfortune happens, the customer has just started using that source, so probably hasn’t built too much stuff on it yet, so asking the customer to recreate the source wouldn’t be disastrous. So I wouldn’t worry too much about this.
2. It might happen that even though the tests catch a regression, but an optimizer change gets badly blocked. The fundamental issue is that many optimizations involve some heuristics, so sometimes it's almost impossible to totally avoid regressions. So maybe there is an optimizer change that leads to a better plan in 99% of cases, but in the remaining 1% it sometimes breaks the MFP guarantee, and it hits one of the active TRANSFORM USING queries, and then this optimizer change is just blocked. I'm a bit struggling to estimate the chances of this happening, but I'm hoping that in these cases it should be possible to keep tweaking the heuristics (with maybe a non-trivial amount of extra work) until the specific query that is occurring in an actual TRANSFORM USING at a customer somehow goes back to being an MFP. @aalexandrov, what do you think?

If we want to further think about 2., then we should find some potentially scary example queries that
- start from an HIR plan that is not obviously an MFP, and
- get optimized down to an MFP after going through non-trivial parts of the optimizer, and
- are non-contrived (i.e., they should look similar to something that a customer might actually type).

Actually, I can't think of even a single example at the moment...

The json example in the design doc is pretty obviously an MFP:

```
materialize=> create table sales_staff(data bytea);
CREATE TABLE
materialize=> explain raw plan for     SELECT
        (data->>'id')::int AS id,
        (data->>'name')::text AS name
    FROM
        (SELECT CONVERT_FROM(data, 'utf8')::jsonb as data FROM sales_staff)
    WHERE
        (data->>'department')::text = 'sales';
                        Raw Plan                         
---------------------------------------------------------
 Project (#1, #2)                                       +
   Map (text_to_integer((#0 ->> "id")), (#0 ->> "name"))+
     Filter ((#0 ->> "department") = "sales")           +
       Project (#1)                                     +
         Map (text_to_jsonb(convert_from(#0, "utf8")))  +
           Get materialize.public.sales_staff           +
```

I also tried some queries with `jsonb_each`, which needs CallTable. This is indeed a bit ugly, but I just realized that it doesn’t end up being an MFP even after optimization, so it’s out of scope, right @petrosagg? (CallTable gets translated to FlatMap, which is not an MFP. Although, I think what TRANSFORM USING wants from MFPs (statelessness, right?) is actually also true for FlatMaps! So, in theory, the scope could be extended to MFPs+FlatMaps, if somebody really wants jsonb_each and co in their TRANSFORM USING. But maybe let’s not do this in the first iteration.)

Reply:
> (raised by Moritz) A customer might create a problematic TRANSFORM USING query between our last test run and the new version going live. This dangerous time interval would be about a day if we run these tests nightly, and could be made even less by an extra last-minute-before-maintenance-window test if this is a big worry. Even with 1 day, the risk is around 2 orders of magnitude less than without the testing (if we assume that a customer has hundreds of days to write a problematic query that does get caught by the test), which might be acceptable. Anyhow, if this misfortune happens, the customer has just started using that source, so probably hasn’t built too much stuff on it yet, so asking the customer to recreate the source wouldn’t be disastrous. So I wouldn’t worry too much about this.

We actually have a design that will allow us to eliminate the window entirely: https://github.com/MaterializeInc/cloud/issues/5057. The idea is that we would lock the catalog (so no new DDL), then validate one last time that all queries in the catalog are still valid, and only _then_ perform the version upgrade. If the customer sneaked in a problematic query last minute, we'd just abort the upgrade for that one environment, and ask the optimizer team to take a look.

> CallTable gets translated to FlatMap, which is not an MFP.

Huh, I don't think I realized that FlapMaps couldn't go in MFPs! Makes sense that it's future work though.